### PR TITLE
New Type class

### PR DESCRIPTION
**Pull request by st-pasha**
The class Type is intended as a replacement for stype/ltype enums.
The main difference from `stype` is that the new class is not an enum, and supports arbitrary type descriptors.

This new class is necessary in order to implement datetime types with various precision / timezone parameters. In the future this class will also be able to accommodate columns of "list" types, compound (struct) columns, union columns, etc.

In the future we want to gradually replace all uses of `stype` with the equivalent code that uses `Type`s.

### CODE REVIEW

**Comment by st-pasha** [comment]
Note: some of the tests are currently failing, but that is addressed by a separate PR #2851 and has nothing to do with this PR.

File: src/core/python/xobject.h [hunk 1]
```diff
@@ -19,6 +19,7 @@
 // FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 // IN THE SOFTWARE.
 //------------------------------------------------------------------------------
+#include <iostream>
```
Comment (oleksiyskononenko) [inline src/core/python/xobject.h:22]: do we need this include?

File: src/core/types/py_type.h [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+//------------------------------------------------------------------------------
+#ifndef dt_TYPES_PY_TYPE_h
+#define dt_TYPES_PY_TYPE_h
+#include "types/type.h"
+#include "python/xobject.h"
+namespace dt {
+
+
+/**
+  * This class corresponds is python-visible `dt.Type`: it describes
```
Comment (oleksiyskononenko) [inline src/core/types/py_type.h:30]: `This class corresponds is python-visible `dt.Type` — typo somewhere?

File: tests/test-types.py [hunk 3]
```diff
@@ -1,12 +1,189 @@
...
+    with pytest.raises(ValueError, match=msg):
+        Type(0)
+    with pytest.raises(ValueError, match=msg):
+        Type(0.5)
+    with pytest.raises(ValueError, match=msg):
+        Type("nothing")
+    with pytest.raises(ValueError, match=msg):
+        Type(type)
+    with pytest.raises(TypeError):
+        Type()
```
Comment (oleksiyskononenko) [inline tests/test-types.py:186]: Add a test accessing `.types` for a frame?

File: src/core/python/xobject.h (hunk 1 above)
Comment (st-pasha) [reply src/core/python/xobject.h:22]: no, this was temporary

File: tests/test-types.py (hunk 3 above)
Comment (st-pasha) [reply tests/test-types.py:186]: Right, forgot about that...

**Review by oleksiyskononenko — APPROVED** [review]

**Pull request merged**